File: packages/hushspec/src/adapters/openai.ts
import type { EvaluationAction, EvaluationResult } from '../evaluate.js';
import type { HushGuard } from '../middleware.js';
import { argsSize } from './tool-mapping.js';

/**
 * Map an OpenAI tool call onto an {@link EvaluationAction}.
 *
 * The call is always a `tool_call` against the function's own name: the API
 * declares no action semantics, and guessing one would consult the wrong rule
 * block. `args_size` records the payload's size without the payload, in the
 * one canonical form core spec 3.7 fixes, so the whitespace a model padded
 * its arguments with does not change the number a `max_args_size` limit is
 * compared against.
 *
 * A model can emit a truncated or malformed JSON string, and an enforcement
 * point that threw on one would fail open: the call would be gated by
 * whatever the caller does with the exception rather than by the policy. Such
 * a payload is measured as received instead, because an unmeasured call is
 * one `max_args_size` cannot bound.
 */
export function mapOpenAIToolCall(
  functionName: string,
  functionArgs: string | Record<string, unknown>,
): EvaluationAction {
  return {
    type: 'tool_call',
    target: functionName,
    args_size: argsSize(functionArgs),
  };
}

export function createOpenAIGuard(
  guard: HushGuard,
): (functionName: string, functionArgs: string | Record<string, unknown>) => EvaluationResult {
  return (
    functionName: string,
    functionArgs: string | Record<string, unknown>,
  ): EvaluationResult => {
    const action = mapOpenAIToolCall(functionName, functionArgs);
    return guard.evaluate(action);
  };
}

File: packages/python/hushspec/adapters/openai.py
from __future__ import annotations

import json
from typing import Callable

from hushspec.evaluate import EvaluationAction, EvaluationResult, args_size_of
from hushspec.middleware import HushGuard


def map_openai_tool_call(
    function_name: str,
    function_args: str | dict,
) -> EvaluationAction:
    """Map one OpenAI tool call onto the action a policy evaluates.

    ``args_size`` is the UTF-8 byte length of the canonical JSON of the
    arguments (core spec 3.7), so a string payload is decoded and measured in
    that one form rather than as the bytes OpenAI happened to send: the model
    may pad the JSON with whitespace or escape a character it need not, and
    neither changes the arguments a ``max_args_size`` rule is about.
    Arguments that are not valid JSON raise: a tool call whose arguments
    cannot be read is not a call a policy should be asked to allow.
    """
    if isinstance(function_args, str):
        function_args = json.loads(function_args)
    args_size = args_size_of(function_args)

    return EvaluationAction(
        type="tool_call",
        target=function_name,
        args_size=args_size,
    )


def create_openai_guard(
    guard: HushGuard,
) -> Callable[..., EvaluationResult]:
    """A handler that maps an OpenAI tool call and evaluates it against *guard*."""

    def handler(
        function_name: str,
        function_args: str | dict,
    ) -> EvaluationResult:
        action = map_openai_tool_call(function_name, function_args)
        return guard.evaluate(action)

    return handler
